src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java

```diff
@@ -1,132 +1,112 @@
-                            .collect(Collectors.joining(", ", "managers {", "}")),
-                    !this.managers.isEmpty() && !this.recipes.isEmpty()? " and " : "",
-                    this.recipes.isEmpty()? "" : this.recipes.stream()
-                            .map(IRecipe::getId)
-                            .map(ResourceLocation::toString)
-                            .collect(Collectors.joining(", ", "recipes {", "}")),
-                    this.rules.stream().map(IReplacementRule::describe).collect(Collectors.joining(", ", "{", "}"))
-            );
-        }
-        
-        @Override
-        public boolean validate(final ILogger logger) {
-            if (this.recipes.isEmpty() && this.managers.isEmpty()) {
-                logger.error("Invalid replacer action: no targeted recipes nor managers");
-                return false;
-            }
-            if (this.rules.isEmpty()) {
-                logger.error("Invalid replacer action: no rules available");
-                return false;
-            }
-            return true;
-        }
-    
-        private Stream<Pair<IRecipeManager, IRecipe<?>>> streamManagers() {
-            return this.managers.stream()
-                    .map(manager -> Pair.of(manager, manager.getAllRecipes().stream().map(WrapperRecipe::getRecipe)))
-                    .flatMap(pair -> pair.getSecond().map(recipe -> Pair.of(pair.getFirst(), recipe)));
-        }
-    
-        private Stream<Pair<IRecipeManager, IRecipe<?>>> streamRecipes() {
-            return this.recipes.stream()
-                    .map(recipe -> Pair.of(RecipeTypeBracketHandler.getCustomManager(Registry.RECIPE_TYPE.getKey(recipe.getType())), recipe));
-        }
-    
-        private <T extends IInventory, U extends IRecipe<T>> Optional<ActionReplaceRecipe> execute(final IRecipeManager manager, final U recipe, final List<IReplacementRule> rules) {
-            try {
-                final IRecipeHandler<U> handler = CraftTweakerRegistry.getHandlerFor(recipe);
-                final Optional<U> newRecipeMaybe = handler.replaceIngredients(manager, recipe, rules);
-            
-                if (newRecipeMaybe.isPresent()) {
-                    return Optional.of(new ActionReplaceRecipe(manager, recipe, newRecipeMaybe.get()));
-                }
-            } catch (final IRecipeHandler.ReplacementNotSupportedException e) {
-                // TODO("Warning maybe?")
-                CraftTweakerAPI.logInfo("Unable to replace ingredients in recipe %s: %s", recipe.getId(), e.getMessage());
-            } catch (final Throwable t) {
-                CraftTweakerAPI.logThrowing("An error has occurred while trying to replace ingredients in recipe %s", t, recipe.getId());
-            }
-            return Optional.empty();
-        }
-    }
+    private static final Supplier<Function<ResourceLocation, ResourceLocation>> DEFAULT_REPLACER = Lazy.concurrentOf(
```

Comment:
Why the 2nd order function?
Like, what advantage does having the Supplier bring instead of using the function directly over here?

Reply:
Originally, this was so the default replacer would be initialized only if needed; with the changes made, though, there is effectively none. I'll remove it